lib/proxy/events/buffer.rb

```diff
@@ -0,0 +1,90 @@
+module Proxy::Events
+
+  LogEvent = Struct.new(:serial, :timestamp, :level, :message)
+  BacktraceLogEvent = Struct.new(:serial, :timestamp, :level, :message, :backtrace)
+
+  class Buffer
+    def self.instance
+      @@buffer ||= Buffer.new
+    end
+
+    # tail buffer only contains warings and errors
+    def initialize(limit = 100, limit_tail = 100, level = ::Logger::Severity::INFO, level_tail = ::Logger::Severity::ERROR)
+      @level = level
+      @level_tail = level_tail
+      @mutex = Mutex.new
+      @main_buffer = RingBuffer.new(limit)
+      @tail_buffer = RingBuffer.new(limit_tail)
+      @serial = 1
+    end
+
+    def push(an_event)
+      @mutex.synchronize do
+        if defined?(an_event.level) && an_event.level >= @level
+          an_event.serial = @serial += 1
+          old_value = @main_buffer.push(an_event)
+          @tail_buffer.push(old_value) if old_value && old_value.level >= @level_tail
+        end
+      end
+    end
+
+    def iterate_ascending
+      @mutex.synchronize do
+        @tail_buffer.iterate_ascending { |x| yield x }
+        @main_buffer.iterate_ascending { |x| yield x }
+      end
+    end
+
+    def iterate_descending
+      @mutex.synchronize do
+        @main_buffer.iterate_descending { |x| yield x }
+        @tail_buffer.iterate_descending { |x| yield x }
+      end
+    end
+  end
+
+  class RingBuffer
+    def initialize(size)
+      @size = size
+      @start = 0
+      @count = 0
+      @buffer = Array.new(size)
+    end
+
+    def full?
+      @count == @size
+    end
+
+    def push(value)
+      stop = (@start + @count) % @size
+      old_value = @buffer[stop]
+      @buffer[stop] = value
+      if full?
+        @start = (@start + 1) % @size
+        old_value
+      else
+        @count += 1
+        nil
+      end
+    end
+    alias :<< :push
+
+    def clear
+      @buffer = Array.new(@size)
+      @start = 0
+      @count = 0
+    end
+
+    def iterate_ascending
+      0.step(@size - 1, 1) do |size_iterator|
+        yield @buffer[(@start + size_iterator) % @size]
+      end
+    end
+
+    def iterate_descending
+      (@size - 1).step(0, -1) do |size_iterator|
+        yield @buffer[(@start + size_iterator) % @size]
+      end
+    end
+  end
```

Comment:
For the record, this class is simplified and slightly enhanced version of https://github.com/celluloid/celluloid/blob/0-16-stable/lib/celluloid/logging/ring_buffer.rb (I added iterators and removed unused methods).